packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -824,9 +825,41 @@ class _TextSelectionHandleOverlayState
         break;
     }
 
+    // On some platforms we may want to calculate the start and end handles
+    // separately so they scale for the selected content.
+    //
+    // For the start handle we compute the rectangles that encompass the range
+    // of the first full selected grapheme cluster at the beginning of the selection.
+    //
+    // For the end handle we compute the rectangles that encompass the range
+    // of the last full selected grapheme cluster at the end of the selection.
+    final InlineSpan span = widget.renderObject.text!;
+    final String text = span.toPlainText();
+    final int firstSelectedGraphemeExtent;
+    final int lastSelectedGraphemeExtent;
+    if(text.isEmpty || widget.selection.isCollapsed){
+      // The call to selectedGraphemes.characters.first/last will throw a state
+      // error if the given string is empty, so fall back to first/last character
+      // range in this case.
+      //
+      // The call to widget.selection.textInside(text) will return a RangeError
+      // for a collapsed selection, fall back to this case when that happens.
+      firstSelectedGraphemeExtent = 1;
+      lastSelectedGraphemeExtent = 1;
+    }else{
+      final String selectedGraphemes = widget.selection.textInside(text);
```

Comment:
This may crash: `widget.renderObject.text` represents the text rendered **last frame**, and `widget.selection` represents the selection to render **this frame**. There's a chance the selection indices are OOB for `text` here.

Reply:
I'd say we stick to either the text & selection from the last frame or the current frame. The former probably makes more sense since we're relying on the text layout metrics here anyway (calling `getRectForComposingRange`). @justinmc what does the text selection menu do in general for something like this?

Reply:
The toolbar is based off of the RenderEditable (see [here](https://github.com/flutter/flutter/blob/e9c1c1627adc424cf3dc78ef3994e457e041503f/packages/flutter/lib/src/widgets/editable_text.dart#L2173)) so I believe everything it's using is based off of the previous frame, and I'd agree that we should do the same here too.

Reply:
Good to know! I agree the text & selection should both come from the same source, I'll make the changes to use the ones from the last frame.